### PR TITLE
Implement model view model and dal for search patient

### PR DESCRIPTION
**Pull request by kmitche9**

### CODE REVIEW

File: CS3230Project/View/Login.resx
Comment (alexofceaser85) [inline CS3230Project/View/Login.resx]: I don't believe this is supposed to be changed

File: CS3230Project/DAL/DataHelper.cs
Comment (alexofceaser85) [inline CS3230Project/DAL/DataHelper.cs]: I prefer that the pattern is a private field this way it is easy to find. Also DataHelper is a weird place to put it. This should be in the model.

File: CS3230Project/DAL/Patients/PatientsDal.cs
Comment (alexofceaser85) [inline CS3230Project/DAL/Patients/PatientsDal.cs]: We should not be adding in the patientId, that is generated by the server

File: CS3230Project/Model/Users/Patients/Patient.cs
Comment (alexofceaser85) [inline CS3230Project/Model/Users/Patients/Patient.cs]: Remove some of this extra whitespace

File: CS3230Project/Model/Users/Patients/PatientManager.cs
Comment (alexofceaser85) [inline CS3230Project/Model/Users/Patients/PatientManager.cs]: In your comments you should avoid the use of > and <. Because C# uses xml like tags for the comment properties (i.e <summary>, <param>, and <returns>) when you add < or > C# thinks that they are opening and closing tags. I prefer uses MORE THAN or LESS THAN instead.

File: CS3230Project/Model/Users/Patients/PatientManager.cs
Comment (alexofceaser85) [inline CS3230Project/Model/Users/Patients/PatientManager.cs]: Should not have have the < and > characters. C# thinks they are opening and closing tags.

File: CS3230Project/Model/Users/Patients/PatientManager.cs
Comment (alexofceaser85) [inline CS3230Project/Model/Users/Patients/PatientManager.cs]: Should not have have the < and > characters. C# thinks they are opening and closing tags.

File: CS3230Project/Model/Users/Patients/Patient.cs
Comment (alexofceaser85) [inline CS3230Project/Model/Users/Patients/Patient.cs:94]: Should not have have the < and > characters. C# thinks they are opening and closing tags.

File: CS3230Project/View/SearchPatient.cs
Comment (alexofceaser85) [inline CS3230Project/View/SearchPatient.cs:1]: The UI still needs some refinement

1. The results table should span the entire screen or stay at the middle of the screen. Currently it is anchored to the left.
2. The date input should be below the Search by Birth Date checkbox. Currently it is at the bottom.
3. The screen should use a footer like the other screens and the back To Home button should be on that footer. (I plan on moving the footer into its own component so you can skip this)
4. The Search button should be in the middle of the screen. Currently it is in the middle of the search by first and last name grid item.
Capture

File: CS3230Project/View/Home.cs
Comment (alexofceaser85) [inline CS3230Project/View/Home.cs:48]: The Search Patients and Register Patient buttons are out of order on the home screen. Currently the Search Patients button is above the Register Patient button. On the UI diagrams the Search Patients button should be below the Register Patient button.

File: CS3230Project/View/Login.resx
Comment (kmitche9) [reply CS3230Project/View/Login.resx]: Fixed.

File: CS3230Project/DAL/DataHelper.cs
Comment (kmitche9) [reply CS3230Project/DAL/DataHelper.cs]: Completed.

File: CS3230Project/DAL/Patients/PatientsDal.cs
Comment (kmitche9) [reply CS3230Project/DAL/Patients/PatientsDal.cs]: Fixed.

File: CS3230Project/Model/Users/Patients/Patient.cs
Comment (kmitche9) [reply CS3230Project/Model/Users/Patients/Patient.cs]: Fixed.

File: CS3230Project/Model/Users/Patients/PatientManager.cs
Comment (kmitche9) [reply CS3230Project/Model/Users/Patients/PatientManager.cs]: Fixed

File: CS3230Project/Model/Users/Patients/PatientManager.cs
Comment (kmitche9) [reply CS3230Project/Model/Users/Patients/PatientManager.cs]: Fixed.

File: CS3230Project/Model/Users/Patients/PatientManager.cs
Comment (kmitche9) [reply CS3230Project/Model/Users/Patients/PatientManager.cs]: Fixed.

File: CS3230Project/Model/Users/Patients/Patient.cs
Comment (kmitche9) [reply CS3230Project/Model/Users/Patients/Patient.cs:94]: fixed.

File: CS3230Project/View/Home.cs
Comment (kmitche9) [reply CS3230Project/View/Home.cs:48]: Fixed.

File: CS3230Project/View/SearchPatient.cs
Comment (kmitche9) [reply CS3230Project/View/SearchPatient.cs:1]: Fixed. Take a look at the revised version and let me know what you think.

**Comment by kmitche9** [comment]
I have fixed all the issues discussed, please review and let me know your thoughts.

File: CS3230Project/View/EditPatient.cs
Comment (alexofceaser85) [inline CS3230Project/View/EditPatient.cs:1]: Not holding up the approval. The edit patient should be in a separate branch as they are separate features and therefore a separate PR, unless I am missing anything. I know the edit patient work is still in progress so I'll approve it although it does need some work.

File: CS3230Project/View/SearchPatient.cs
Comment (alexofceaser85) [inline CS3230Project/View/SearchPatient.cs]: Is holding up approval. Sorry I did not catch this on my first look, but you do need try catch blocks around this code. When the user clicks the search patient button with empty fields the application crashes because it does not have try catch blocks. So basically the whole method should be wrapped in try catch blocks like this.

            try
            {
                List<Patient> matchingPatients = new List<Patient>();
                if (this.searchFirstAndLastNameCheckBox.Checked && this.searchByBirthDateCheckBox.Checked)
                {
                    matchingPatients = PatientManager.GetPatientsByNameAndDateOfBirth(this.firstNameTextBox.Text,
                        this.lastNameTextBox.Text, this.dateOfBirthDatePicker.Value);
                }
                else if (this.searchFirstAndLastNameCheckBox.Checked)
                {
                    matchingPatients =
                        PatientManager.GetPatientsByName(this.firstNameTextBox.Text, this.lastNameTextBox.Text);
                }
                else if (this.searchByBirthDateCheckBox.Checked)
                {
                    matchingPatients = PatientManager.GetPatientsByDateOfBirth(this.dateOfBirthDatePicker.Value);
                }

                this.displayPatientData(matchingPatients);
            }
            catch (ArgumentException exception)
            {
                MessageBox.Show(exception.Message, "SAMPLE ERROR");
            }

File: CS3230Project/View/SearchPatient.cs
Comment (alexofceaser85) [inline CS3230Project/View/SearchPatient.cs:2]: Not holding up Approval. At some point we might want to make the search by name checked on by default. Would like your opinion on this. Again this is just a comment and not holding up the Approval.

File: CS3230Project/Model/DataValidator.cs
Comment (alexofceaser85) [inline CS3230Project/Model/DataValidator.cs:26]: I would consider moving the null check back into the patient class. But I do like having a separate class for common regex validation.

File: CS3230Project/Model/Users/Patients/PatientManager.cs
Comment (alexofceaser85) [inline CS3230Project/Model/Users/Patients/PatientManager.cs:29]: I would separate out the precondition checks. From a users point of view they don't know what null really is. So when they see a message that something is null or empty it can be a bit confusion. I also like my error messages to be as clear as possible and point exactly to what is wrong. Does not have to hold up the approval but a separate ticket would need to be made if this is something that is easier for us to work on later.

Comment (alexofceaser85) [reply CS3230Project/Model/Users/Patients/PatientManager.cs:29]: So I would throw separate argument exceptions for null and for empty. With separate error messages.

File: CS3230Project/Model/Users/Patients/PatientManager.cs
Comment (alexofceaser85) [inline CS3230Project/Model/Users/Patients/PatientManager.cs:83]: Would prefer if these precondition checks were separated out

**Review by alexofceaser85 — CHANGES_REQUESTED** [review]

File: CS3230Project/View/EditPatient.cs
Comment (kmitche9) [reply CS3230Project/View/EditPatient.cs:1]: That was an accident on my part.

File: CS3230Project/View/SearchPatient.cs
Comment (kmitche9) [reply CS3230Project/View/SearchPatient.cs]: Fixed. Good catch!

File: CS3230Project/View/SearchPatient.cs
Comment (kmitche9) [reply CS3230Project/View/SearchPatient.cs:2]: Agreed! It gets annoying while testing. It has been fixed.

**Comment by kmitche9** [comment]
Main issue fixed.

File: CS3230Project/ErrorMessages/ErrorMessages.cs
Comment (alexofceaser85) [inline CS3230Project/ErrorMessages/ErrorMessages.cs]: That is a start but again there should be separate checks for null and empty

File: CS3230Project/Model/Users/Patients/PatientManager.cs
Comment (alexofceaser85) [inline CS3230Project/Model/Users/Patients/PatientManager.cs:32]: Checks should be seperate

**Review by alexofceaser85 — APPROVED** [review]

File: CS3230Project/ErrorMessages/ErrorMessages.cs
Comment (kmitche9) [reply CS3230Project/ErrorMessages/ErrorMessages.cs]: That was changed because I changed the error message that is display in the window when searching by empty names. I have not attempted to fix this yet.

**Pull request merged**
